[자동차 경주 구현] 해나 미션 제출합니다.

hxeyexn opened the pull request:
안녕하세요 리뷰어님~

안드로이드 6기 2팀 해나입니다!
많이 부족하지만 잘 부탁드리겠습니다~

**🧐 질문**
미션을 진행하면서 궁금한 점이 있어 질문드립니다!

1.
``` kotlin
    private fun getNames(): List<String> {
        outputView.printCarNamesGuide()
        lateinit var names: List<String>

        while (true) {
            try {
                names = inputView.readCarNames().split(COMMA)
                validateNames(names)
                names.forEach { name ->
                    Car(name)
                }
                break
            } catch (e: IllegalArgumentException) {
                println(e.message)
            }
        }

        return names
    }
```
 - Controller로 Race class 를 사용하고 있습니다. 이때, getNames() 메서드에서 OutputView를 통해 입력가이드문을 출력하는 부분이 있는데(위 코드 2번째 줄), 이러한 출력부를 별도의 메서드로 분리하는 것이 좋은지, 아니면 메소드명을 수정하는 것이 좋은지 알고 싶습니다!

2.  입력값 검증(Validation)을 InputView 에서 하는게 좋은지, 아니면 model 혹은 controller 에서 진행하는 것이 좋은지 알고 싶습니다.
 (이번 미션에서는 model과 controller에서 유효성 검증을 했는데, depth가 깊어지는 문제가 있었습니다…)

Hogu59 committed:
docs: Add requirement in README.md

Add fundamental requirements.
Add requirements for exception handling (Car, RoundNumber)

Hogu59 committed:
feat(view): Implement print car name input guide

view package created.
Printing Car name input Guide announcement function is added to OutputView.kt

Hogu59 committed:
docs: modify readme(outputview - car name request)

Readme file modified. OutputView -> Car name input Guide announcement printed correctly.

Hogu59 committed:
feat(view): implement car names input function

Function getting input of car names is added to InputView.
Return type is String, and the validation would be conducted in Controller and Model.

Hogu59 committed:
feat(view): implement print round number guide

Printing round number input guide announcement function added.

Hogu59 committed:
feat(view): implement input round number function

Function getting input of round number is added to InputView.
Return type is String, and the validation would be conducted in Controller and Model.

Hogu59 committed:
docs: modify Readme(car name, round number)

Readme file modified.
input and output functions related with car name and round number are functioning correctly.

Hogu59 committed:
fix(view): Add constant for round number guide

Constant for Round Number input announcement added.

Hogu59 committed:
feat(view): implement print result header

Function for printing execution result header added.

Hogu59 committed:
docs: modify readme(outputview - result header)

Readme file modified. OutputView -> result header printed correctly.

Hogu59 committed:
feat(model): implement car class

Car (data) class added for model.

Hogu59 committed:
feat(controller): implement get name list function

Controller package added.
Race class added.
getNames function, makes string to List<String> from inputView data, added.

Hogu59 committed:
feat(controller): implement validation nameList

Function for validating name list added.

Hogu59 committed:
feat(model): validate car name in model

Function for car name validation added.
Exceptions,
1) Empty input
2) name with space

Hogu59 committed:
feat(model): generate random number

Function for generating random number added.

Hogu59 committed:
feat(model): determine when to move

Function for determining when to move added.
If random number is same or greater than 4, the position would be changed)
(as, move was conducted)

Hogu59 committed:
feat(controller): implement getting round number

Function getting round number added.
Get string input from InputView, and return int.

Hogu59 committed:
feat(controller): validate round number

Function for validating round number input added.
Exceptions,
1) input is out of int range
2) input is not number (ex. letter, etc.)
3) input is empty

Hogu59 committed:
feat(controller): implement getting car list

Function returning List<Car> from List<string> (names) added.

Hogu59 committed:
docs: modify readme(determine move)

Function determining to move is working properly.

Hogu59 committed:
Feat(controller): execute all rounds

Start function added.
Function for executing all rounds added.

Hogu59 committed:
feat(view): print results for each round

Function to print current status(name, position) was added.

Hogu59 committed:
feat(view): print new line

Function to print new line added.

Hogu59 committed:
feat(controller): print results for all rounds

Implement printing current car status. (name, position)

Hogu59 committed:
docs: modify readme(print result for each round)

result for each round printed correctly.

Hogu59 committed:
feat(controller): get winner list

Function getting winner list added.
Get winner list, list of strings(name).

Hogu59 committed:
feat(view): print winners

Function for printing winner list added.

Hogu59 committed:
feat(controller): show winners

Function to print winner list added in Controller.

Hogu59 committed:
feat(controller): use functions for result

Function related with result printing called in start function.

Hogu59 committed:
feat(application): call start function in main

Call start function for Controller in main function of Application.

Hogu59 committed:
test: test for car name exceptions

Exception checked,
1) empty name input
2) name length 0 or more than 5
3) name with space

Hogu59 committed:
test: test for race class(controller)

Exception checked,
1) empty name list
2) size of name list is less than 2
3) size of name list is greater than 20
4) when duplicated name exist
5) when round number input exceeds int range
6) when round number input is letter of string
7) when round number input is empty
8) when round number input is less than 1

Hogu59 committed:
test: test for printing winner

Test conducted to check getting right winner.

Hogu59 committed:
feat(dependency): use junit5 explicitly

Modification of codes to use junit5 explicitly.

Hogu59 committed:
docs: modify readme

All requirements were checked, including exceptions.

Hogu59 committed:
refactor(controller): seperate constants

Move the values of hard-coding to companion object.
To make more efficient to re-use.

Hogu59 committed:
refactor(model): seperate constants

Move the values of hard-coding to companion object.
To make more efficient for further modification.

Hogu59 committed:
refactor(controller): seperate functions

Function seperated to make depth shallower.

Hogu59 committed:
refactor(controller): modify error message

Error message divided by each cases.

Hogu59 committed:
refactor(model): modify error message

Error message divided by each cases.

Hogu59 committed:
refactor(controller): make functions private

Functions only used in local are changed private.

Hogu59 committed:
refactor(model): make functions private

Functions only used in local are changed private.

ghojeong commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private fun getNames(): List<String> {
+        outputView.printCarNamesGuide()
```
> Controller로 Race class 를 사용하고 있습니다. 이때, getNames() 메서드에서 OutputView를 통해 입력가이드문을 출력하는 부분이 있는데, 이러한 출력부를 별도의 메서드로 분리하는 것이 좋은지, 아니면 메소드명을 수정하는 것이 좋은지 알고 싶습니다!

메소드명을 수정한다는 말은 `getNames` 를 `getNamesAfterPrintCarNamesGuide` 처럼 메서드 명으로 입력 가이드 문이 출력된다는 점을 표현할 수 있도록 한다는 말인가요?

출력부를 별도의 메서드로 분리한다는 것은 무슨 로직을 어떤 메서드명으로 분리하겠다는 것인지 질문의 의도를 잘 모르겠습니다.
아래처럼 메서드를 하나 더 만드시려는 계획 이신가요?

```kt
    private fun printCarNamesGuide() = outputView.printCarNamesGuide()
```

무엇이 좋은지를 알려하기 보다는, 해나님께서 느끼시기에 좋지 않다고 생각하는 부분을 말씀해주시면 좀 더 같이 생각해 볼 수 있을 것 같습니다.
getNames() 라는 메서드 안에 있는 `outputView.printCarNamesGuide()` 가 왜 적절치 않다고 생각하는지, 무슨 나쁜 점이 있길래 분리 혹은 수정하려 하는 것인지 표현해 주시면 좋을 것 같습니다.

hxeyexn replied:
메서드명을 수정한다는 말은 입력 가이드 문이 출력된다는 점을 표현할 수 있도록 한다는 말이 맞습니다! 메서드를 분리한다는 것은 입력 가이드를 보여주는 부분과 자동차의 이름을 가져오는 로직이 한 메서드에 있어서 가이드를 보여주는 메서드를 별도로 만들려는 계획이었습니다. 

미션을 함께한 페어는`outputView.printCarNamesGuide()`도 차량 이름을 가져오기 위한 로직이니 한 메서드로 묶는 방향으로 작성하자고 제시했습니다. 페어 말 처럼 하나의 기능이라고 묶을 수도 있지만, 저는 `outputView.printCarNamesGuide()`(입력 가이드를 보여주는 부분)과 자동차 이름을 가져오는 로직은 별도의 기능이라고 생각합니다.  한 메서드는 하나의 기능만 가지고 있어야 한다고 생각했기에 getNames() 안에 `outputView.printCarNamesGuide()` 적절치 않다고 생각했습니다. 기능을 얼마나 세분화할 것 인가의 차이인 것 같은데 어느 정도까지 메서드를 세분화해야할 지 고민이 되어 여쭤본 질문이었습니다.

ghojeong replied:
> 기능을 얼마나 세분화할 것 인가의 차이인 것 같은데 어느 정도까지 메서드를 세분화해야할 지 고민이 되어 여쭤본 질문이었습니다.

어느 정도까지 세분화해야하는지 고민이 되신다면,
해나님이 해보실 수 있는 최대한도로 세분화하는 일에 도전해보세요.

메서드를 세분화한 후에는, 그 세분화한 메서드를 모아서 실행하는 상위 메서드가 새롭게 필요하게 됩니다.
그렇게 메서드간에 생기는 계층이 abstraction 이고, 그 abstraction 의 정도에 따라 abstraction level 이 생기게 됩니다.

세분화한 메서드를 실행해주는 메서드를 합치는 메서드를 다시 선언하다보면,
이건 너무 세분화한 것 같아서 그냥 하나의 메서드로 합치고 싶은 생각도 들 것입니다.
그때는 그냥 합치시면 됩니다.

처음에 최대한도로 세분화하라는 조언은, 보통 합쳐진것을 쪼개는 리팩토링 보다,
쪼개진 것을 합치는 리팩토링이 더 쉽기 때문입니다.
그래서 저는 처음에 일부러 많이 쪼개놓고서, 나중에 다시 합치고는 합니다.

이 작업을 많이 반복하다보면, 어느 정도까지 세분화해야하는지 본인만의 기준이 생깁니다.
너무 세분화했을 때의 불편함을 몸소 겪게되어, 이 이상은 세분화하지 말아야겠다고 설명할 수 있는 개인 경험이 생기기 때문입니다.

hxeyexn replied:
항상 어느 정도 세분화해야 할지 고민이었는데 파이로님 조언 덕분에 어느 정도 가닥이 잡힌 것 같습니다! 처음에 최대한 세분화하는 과정을 반복하면서 저만의 세분화 기준을 찾아보겠습니다. 감사합니다 🙇

ghojeong commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,152 @@
+        outputView.printCarNamesGuide()
+        lateinit var names: List<String>
+
+        while (true) {
+            try {
+                names = inputView.readCarNames().split(COMMA)
+                validateNames(names)
+                names.forEach { name ->
+                    Car(name)
+                }
+                break
+            } catch (e: IllegalArgumentException) {
+                println(e.message)
+            }
+        }
+
+        return names
+    }
+
+    fun validateNames(names: List<String>) {
+        require(names.isNotEmpty()) { ERROR_REQUEST_INPUT }
+        require(names.size >= MIN_NAMES_SIZE) { ERROR_CAR_LESS_THAN_TWO }
+        require(names.size <= MAX_NAMES_SIZE) { ERROR_CAR_GREATER_THAN_TWENTY }
+        require(names.distinct().size == names.size) { ERROR_NAME_DUPLICATION }
```
> 입력값 검증(Validation)을 InputView 에서 하는게 좋은지, 아니면 model 혹은 controller 에서 진행하는 것이 좋은지 알고 싶습니다.
(이번 미션에서는 model과 controller에서 유효성 검증을 했는데, depth가 깊어지는 문제가 있었습니다…)

입력값 검증에 대한 테스트 코드를 작성하고자 한다면, model 혹은 controller 에서 진행하는게 좋아보입니다.
반면 입력값 검증이 핵심 비즈니스 로직이 아니라 테스트가 굳이 필요 없다면 view 에서 진행하는 것도 괜찮습니다.

지금은 검증 로직에 대한 테스트를 일일히 전부 작성해주셨는데,
view 로 옮기게 된다면 이 테스트 코드를 어떻게 리팩토링하려 하시는지 궁금합니다.

또한 depth 가 깊어지는 문제는 유효성 검증과는 상관없이,
한 메서드 안에서 이중 루프가 돌게되어 생기게 되는 문제로 생각되는데 어떻게 생각하시나요?

hxeyexn replied:
이번 자동차 경주 미션에서는 특정 조건을 만족해야만 처리를 허용하는 부분이 많기에 입력값 검증도 핵심 비즈니스 로직이라 생각되어 model 혹은 controller에서 하는 게 좋겠다는 생각이 드네요!

view로 검증 로직을 옮기게 된다면 검증 로직 관련 메서드를 validation이라는 별도의 class로 옮기고, 이를 inputview에서 가져다 쓸 생각이어서 테스트 코드는 별도의 validationTest를 만들어 현재 코드와 동일하게 검증 로직에 대한 테스트를 일일히 진행하는 방향으로 리팩토링할 생각이었습니다.

Car model에도 이름에 대한 입력값 검증 코드가 있어 depth가 깊어지는 게 아닌가 생각했습니다. 또한 while 안에 try-catch 문이 돌고 있어 break로 인해 메서드를 따로 분리하지 못하다보니 depth가 더 깊어지는 문제가 발생했다고 생각합니다. 예외를 던져줄 때 재귀방식으로 변경해서 depth를 줄이면 될 지 아니면 다른 방법이 있는지 궁금합니다!

ghojeong replied:
try-catch 를 쓰면 문법 때문에 indentation 의 깊이가 깊어져서, depth 가 깊어져 보일 수 있지만,
사실 크게 상관없습니다.

depth 를 이야기할 때, 쉬운 설명을 위해서 indentation 을 이야기하지만,
엄밀히 이야기하면 abstraction level 을 뜻합니다.
SOLID 원칙 중 Single Responsibility 원칙은 한 함수가 하나의 목적을 위해 동작하기를 권장하는데,
이는 abstraction level 이 2가 넘지 않도록 한 함수내의 로직을 추상화 하라는 의미도 있습니다.

재귀 함수나 알고리즘을 통한 리팩토링을 하려 하기 보다는,
하나의 함수가 하나 이상의 역할을 하고 있다면 그 함수를 더 작게 쪼개는 연습을 해보셨으면 합니다.

가령 지금의 getNames 의 경우
1. 사용자에게 문자열을 읽어오기
2. 읽어온 문자열을 이름으로 쪼개기
3. 쪼개진 이름들을 검증하기
4. 검증된 이름으로 Car 객체 만들기

이렇게 4가지 작업을 하고 있습니다.

제가 동료 개발자라고 가정하고 코드를 넘겨받았을 때, `fun getNames(): List<String>` 라는 함수명과 인터페이스로 
유추할 수 있는 동작은 "1. 어딘가에서 문자열을 가져와서, 2. 그 문자열로 이름 리스트를 만드나 보다" 정도입니다.
그 보다 깊은 레벨의 동작이나 그 보다 넓은 의미의 동작은 추측할 수 가 없습니다.

해나 님이 알고 있는 개발자 파트너에게, 함수명과 입력타입, 반환 타입만 알려주면,
그 파트너가 해나님이 지금 구현하신 것과 동일한 기능을 빠짐없이 구현할 수 있도록 함수를 작고 직관적으로 쪼개주세요.

조직에서 함께 일할 때 중요한 점은, 내가 마음 편하게 휴가를 가더라도, 동료에게 자신의 일을 믿고 맡길 수 있어야 한다는 점입니다.
심지어 동료에게 넘겨준 제 코드에 제대로 동작하지 않는 부분이 있다고 하더라도 말입니다.

지금과 같은 상황이라면 해나님이 휴가를 가셨고 무언가 버그가 생겼을 때, 
동료인 제가 getNames 메서드에서 왜 Car 객체를 만들고 있는지 물어보기 위해 전화를 하는 경우가 생길 것 같습니다.
업계에 떠도는 맨날 야근하는 개발 조직 괴담이 역할과 책임이 명확하지 않은 함수, 클래스, 혹은 모듈로 인해서 생긴다고 생각해주세요. ㅎㅎ 🌃

hxeyexn replied:
getNames가 메서드명과 관련없는 작업을 많이 하고 있었네요.. 피드백 반영하여 getNames, getRoundNumber 메서드를 최대한 직관적으로 쪼개보았습니다. 이에 대해 한번 더 피드백 주시면 감사하겠습니다!
예시를 상세히 설명해주셔서 이해가 무척 잘 됐습니다! 감사합니다 👍

ghojeong commented on README.md:
```diff
@@ -1 +1,27 @@
\ No newline at end of file
+# kotlin-racingcar
+
+## 🏎 ️자동차 경주
+
+ - [x] 자동차 입력 문구 출력
+ - [x] 자동차 이름 입력
+ - [x] 시도 횟수 입력 문구 출력
+ - [x] 시도 횟수 입력
+ - [x] 실행 결과 헤더 출력
```
요구사항을 README 에 예쁘게 적어주셨군요 👍 
앞으로도 계속 README 를 성실히 적어주시면 감사하겠습니다.

hxeyexn replied:
감사합니다 😄

ghojeong commented on src/test/kotlin/racingcar/RaceTest.kt:
```diff
@@ -0,0 +1,125 @@
+import org.junit.jupiter.params.provider.ValueSource
+import racingcar.controller.Race
+
+class RaceTest {
+    private val race = Race()
+
+    @Test
+    fun inputNone() {
+        val exception =
+            assertThrows<IllegalArgumentException> {
+                race.validateNames(emptyList())
+            }
+
+        assertEquals(
+            ERROR_MESSAGE,
+            exception.message,
+        )
+    }
```
현재 RaceTest 안에 있는 테스트가 전혀 통과를 못하고 있는 듯 한데 확인 부탁드립니다. 😭

hxeyexn replied:
확인해보니 테스트 구현 후 리팩토링 과정에서 유효성 검사 error message를 세분화 했는데, 테스트 코드는 깜빡하고 수정하지 않아서 RaceTest 안에 있는 테스트가 하나도 통과가 안 되네요😭 리팩토링하고 테스트 코드도 다시 돌려봤어야 하는 데 실수가 있었네요..! 해당 부분 수정해서 다시 리뷰 요청 드리겠습니다!

ghojeong commented on src/test/kotlin/racingcar/CarTest.kt:
```diff
@@ -0,0 +1,31 @@
+import racingcar.model.Car
+
+class CarTest {
+    @Test
+    fun inputIsEmpty() {
+        assertThrows<IllegalArgumentException> {
+            Car("")
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = ["", "crongcrong"])
+    fun outOfRangeNameLength(name: String) {
+        assertThrows<IllegalArgumentException> {
+            Car(name)
+        }
+    }
+
+    @Test
+    fun inputContainsSpace() {
+        assertThrows<IllegalArgumentException> {
+            Car("he na")
+        }
```
테스트 의도를 영어 메서드명 만으로 표현하는 것도 좋지만, DisplayName 어노테이션을 활용하면,
테스트에 관한 더 자세한 설명을 적을 수 있습니다.
한국인 동료와 같이 협업한다고 생각하고, 테스트에 관한 설명을 자세히 적어주시면 감사하겠습니다.

hxeyexn replied:
DisplayName 어노테이션은 처음 들어봤는데, 활용하면 확실히 동료들이 해당 테스트가 어떤 동작을 하고 있는지 이해하기 좋을 것 같네요!

ghojeong commented on src/test/kotlin/racingcar/WinnerTest.kt:
```diff
@@ -0,0 +1,24 @@
+import racingcar.controller.Race
+import racingcar.model.Car
+
+class WinnerTest {
+    private val race = Race()
+
+    @Test
+    fun checkWinners() {
+        val cars =
+            listOf(
+                Car("crong", 5),
+                Car("hena", 2),
+                Car("pang", 3),
+            )
+        val expectedResult = listOf("crong")
+        val winners = race.getWinners(cars)
```
WinnerTest 를 돌리니, 아래와 같은 에러 메시지가 뜹니다. 😭 
해결 부탁드립니다.

<img width="753" alt="스크린샷 2024-02-16 오전 1 17 48" src="https://github.com/woowacourse/kotlin-racingcar/assets/38595650/13cf914a-aa60-4300-ad42-82d487c0280f">

hxeyexn replied:
해당 부분도 리팩토링 후 테스트 코드를 실행해보지 않아 생기는 오류네요 🫠 해결해서 다시 리뷰 올리도록 하겠습니다!

ghojeong reviewed (changes requested):
해나님!
자동차 경주 미션을 함께 하게된 파이로 리뷰어입니다! 🙇 
본격적인 리뷰를 하기에 앞서, 테스트를 실행해보니,
통과하지 못하는 테스트가 있음을 발견했습니다.
수정하시고 다시 리뷰 요청주시면 감사하겠습니다!

hxeyexn committed:
fix(racetest): 테스트가 통과되지 않는 오류 수정

리팩토링 과정에서 유효성 검사 에러 메세지 세분화 후 테스트 코드를 수정하지 않아 발생하는 오류 해결

hxeyexn committed:
fix(Race): WinnerTest가 실행되지 않는 오류 해결

maxPosition 별도 메서드로 분리, 멤버 변수 cars를 지역변수 cars로 변경

hxeyexn committed:
refactor(race): getNames 메서드 직관적으로 세분화

1. 자동차 이름 입력 가이드 출력 메서드

2. 사용자에게 문자열을 읽어오는 메서드

3. 읽어온 문자열을 이름으로 쪼개는 메서드

4. 쪼개진 이름들을 검증하는 메서드

5. 검증된 이름으로 Car 객체 만드는 메서드

6. Cars를 가져오는 메서드

hxeyexn committed:
refactor(race): getRoundNumber 메서드 직관적으로 세분화

1. 시도 횟수 입력 가이드 출력 메서드

2. 사용자에게 문자열을 읽어오는 메서드

3. RoundNumber 가져오는 메서드

hxeyexn committed:
refactor(Race): 에러 메세지 출력문 outputview로 이동

hxeyexn committed:
refactor(cartest): 테스트 코드에 대한 상세 설명 추가

DisplayName 어노테이션을 활용해 테스트 코드에 대한 상세 설명 추가

hxeyexn committed:
refactor(racetest): 테스트 코드에 대한 상세 설명 추가

DisplayName 어노테이션을 활용해 테스트 코드에 대한 상세 설명 추가

hxeyexn committed:
refactor(winnertest): 테스트 코드에 대한 상세 설명 추가

DisplayName 어노테이션을 활용해 테스트 코드에 대한 상세 설명 추가

ghojeong commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,151 @@
+
+    private fun showCarNamesGuide() = outputView.printCarNamesGuide()
+
+    private fun getNames(): String = inputView.readCarNames()
+
+    private fun splitNames(): List<String> = getNames().split(COMMA)
+
+    fun validateNames(names: List<String>) {
+        require(names.isNotEmpty()) { ERROR_REQUEST_INPUT }
+        require(names.size >= MIN_NAMES_SIZE) { ERROR_CAR_LESS_THAN_TWO }
+        require(names.size <= MAX_NAMES_SIZE) { ERROR_CAR_GREATER_THAN_TWENTY }
+        require(names.distinct().size == names.size) { ERROR_NAME_DUPLICATION }
+    }
+
+    private fun makeCars(): List<Car> {
+        val names = splitNames()
+        val cars = mutableListOf<Car>()
+
+        validateNames(names)
+        names.forEach { name ->
+            cars.add(Car(name))
+        }
+        return cars
+    }
```
함수 쪼개 주신 것 무척 마음에 듭니다. 👍 
이 정도라면 내부 로직을 읽지 않고, 함수명과 인터페이스만 보고서도 이 함수가 무슨 동작을 할 수 있는지 알 수 있을 것 같습니다.

사람마다 저마다의 클린 코드 기준이 있겠지만, 저 같은 경우는 **나와 내 동료의 할일이 얼마나 줄어드는가?** 를 제일 중요하게 봅니다.
제 기준에서 코드를 클린하게 작성했다면, 저와 제 동료가 읽어야할 코드의 양이 줄어들어야 합니다. (물리적인 코드 줄 수가 작다는 의미가 아닙니다.)

파이로 님이라면 믿고 코드를 스킵해서 읽어도 되지!!!! 오늘은 일찍 퇴근합시다!!!!
하는 느낌을 동료에게 줄 수 있을 때 저는 제일 뿌듯하더라고요.

ghojeong commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    fun getValidRoundNumber(roundNumberInput: String): Int {
+        require(checkRoundNumber(roundNumberInput)) { ERROR_INPUT_NUMBER_ONLY }
+        require(roundNumberInput.isNotEmpty()) { ERROR_REQUEST_INPUT }
+
+        val roundNumber = roundNumberInput.toInt()
+        require(roundNumber >= MIN_ROUND_NUMBER) { ERROR_ROUND_LESS_THAN_ONE }
+
+        return roundNumber
+    }
+
+    fun checkRoundNumber(roundNumber: String): Boolean {
+        try {
+            roundNumber.toInt()
+            return true
+        } catch (e: NumberFormatException) {
+            throw IllegalArgumentException(ERROR_INPUT_NUMBER_ONLY)
+        }
+    }
+
+    private fun showResult(roundNumber: Int) {
+        outputView.printNewLine()
+        outputView.printResultHeader()
+
+        repeat(roundNumber) {
+            showRaceStatus()
+            outputView.printNewLine()
+        }
+    }
+
+    private fun showRaceStatus() {
+        cars.forEach { car ->
+            car.move()
+            outputView.printRaceResult(car.name, car.position)
+        }
+    }
+
+    fun getWinners(cars: List<Car>): List<String> {
```
step1 에서 함수를 쪼개는 것을 연습해 보았다면,
step2 에서는 클래스(파일)를 쪼개는 연습을 해보셨으면 합니다.

현재 Race 가 테스트가 필요할 정도로 굉장히 많은 로직들이 있는데,
"테스트가 필요한 로직이 있다" 라고 한다면 모두 model 이라고 생각해주세요.

Controller 에는 start() 메서드만 public 하게 공개하고,
나머지는 별도의 model 클래스를 선언해서 옮겨 주시기 바랍니다.

잘 쪼개진 함수는 동료가 스킵해서 읽을 수 있는 것처럼,
잘 쪼개진 클래스 또한 동료가 스킵해서 읽을 수 있습니다.

저는 보통 클래스에 IDE 에서 점을 찍으면 자동완성으로 보이는,
public 메서드만 읽고 해당 클래스 코드를 읽는 것을 스킵하고는 합니다.

아마 해나님도 외부 라이브러리를 쓰실 때, 저와 비슷하게 자동완성으로 제공되는 public 함수만 읽어보지,
실제 내부 코드와 로직을 까서 읽어보는 경우는 거의 없으시리라 생각됩니다.

ghojeong commented on src/test/kotlin/racingcar/CarTest.kt:
```diff
@@ -0,0 +1,35 @@
+import org.junit.jupiter.params.provider.ValueSource
+import racingcar.model.Car
+
+class CarTest {
+    @DisplayName("자동차 이름을 작성하지 않았을 때")
+    @Test
+    fun inputIsEmpty() {
+        assertThrows<IllegalArgumentException> {
+            Car("")
+        }
+    }
+
+    @DisplayName("자동차 이름 길이가 1자 미만 5자 초과일 때")
+    @ParameterizedTest
+    @ValueSource(strings = ["", "crongcrong"])
+    fun outOfRangeNameLength(name: String) {
+        assertThrows<IllegalArgumentException> {
+            Car(name)
+        }
+    }
+
+    @DisplayName("자동차 이름 안에 공백이 있는 경우")
+    @Test
+    fun inputContainsSpace() {
+        assertThrows<IllegalArgumentException> {
+            Car("he na")
+        }
```
<img width="573" alt="스크린샷 2024-02-17 오전 9 40 43" src="https://github.com/woowacourse/kotlin-racingcar/assets/38595650/87e01754-3220-4b6b-a187-1e905ebcb789">

그렇게 model 로 옮겨진 클래스들은 method 의 테스트 커버리지가 100% 가 되도록 해주세요.
model 로 옮겼다는 것은 public method 를 전부 테스트하고 검증받겠다는 강력한 의지의 표명이라고 생각하셔도 됩니다. ㅋㅋㅋㅋ

Line Coverage 는 욕심 안 내셔도 괜찮습니다.

ghojeong reviewed (approved):
해나님!
메서드를 쪼개달라는 step1 피드백 매우 훌륭하게 반영해주셨습니다. 👍 💯 
step2 에서는 객체를 쪼개는 일에 도전해보시기 바랍니다.
step2 에서 도움이 될 만한 도전과제들을 피드백으로 남겨보았습니다.
step2 도 화이팅입니다! 🔥

ghojeong reviewed (approved):
해나님!
메서드를 쪼개달라는 step1 피드백 매우 훌륭하게 반영해주셨습니다. 👍 💯 
step2 에서는 객체를 쪼개는 일에 도전해보시기 바랍니다.
step2 에서 도움이 될 만한 도전과제들을 피드백으로 남겨보았습니다.
step2 도 화이팅입니다! 🔥
